// crass/dist/nodes/ElementSelector.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

module.exports = function () {
    /**
     * @constructor
     * @param {string} ident
     * @param {string} ns
     */
    function ElementSelector(ident, ns) {
        _classCallCheck(this, ElementSelector);

        this.ident = ident;
        this.ns = ns;
    }

    /**
     * @return {string}
     */


    _createClass(ElementSelector, [{
        key: 'toString',
        value: function toString() {
            if (this.ident && this.ns) {
                return this.ident + '|' + this.ns;
            } else if (this.ns) {
                return '|' + this.ns;
            } else {
                return this.ident;
            }
        }

        /**
         * @return {string}
         */

    }, {
        key: 'pretty',
        value: function pretty() {
            return this.toString();
        }

        /**
         * @return {ElementSelector}
         */

    }, {
        key: 'optimize',
        value: function optimize() {
            // OPT: Lowercase element names.
            this.ident = this.ident.toLowerCase();
            if (this.ns) {
                this.ns = this.ns.toLowerCase();
            }
            return this;
        }
    }]);

    return ElementSelector;
}();
