// crass/dist/nodes/Import.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

module.exports = function () {
    /**
     * @constructor
     * @param {URL} href
     * @param {MediaList} media
     */
    function Import(href, media) {
        _classCallCheck(this, Import);

        this.href = href;
        this.media = media;
    }

    /**
     * @return {string}
     */


    _createClass(Import, [{
        key: 'toString',
        value: function toString() {
            return '@import ' + this.href.asString() + (this.media ? ' ' + this.media.toString() : '') + ';';
        }

        /**
         * @return {string}
         */

    }, {
        key: 'pretty',
        value: function pretty() {
            return this.toString() + '\n';
        }

        /**
         * @return {Import}
         */

    }, {
        key: 'optimize',
        value: function optimize() {
            return this;
        }
    }]);

    return Import;
}();
