// crass/dist/nodes/Keyframe.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var optimization = require('../optimization');
var utils = require('../utils');

module.exports = function () {
    /**
     * @constructor
     * @param {string} stop
     * @param {*} content
     */
    function Keyframe(stop, content) {
        _classCallCheck(this, Keyframe);

        this.stop = stop;
        this.content = content;
    }

    /**
     * @return {string}
     */


    _createClass(Keyframe, [{
        key: 'toString',
        value: function toString() {
            return utils.joinAll(this.stop, ',') + '{' + this.toStringBody() + '}';
        }

        /**
         * @return {string}
         */

    }, {
        key: 'toStringBody',
        value: function toStringBody() {
            return utils.joinAll(this.content, ';');
        }

        /**
         * @param {int} indent
         * @return {string}
         */

    }, {
        key: 'pretty',
        value: function pretty(indent) {
            var output = '';
            output += utils.indent(utils.joinAll(this.stop, ', ', function (x) {
                return x.pretty(indent);
            }) + ' {', indent) + '\n';
            output += this.content.map(function (line) {
                return utils.indent(line.pretty(indent + 1) + ';', indent + 1);
            }).join('\n') + '\n';
            output += utils.indent('}', indent) + '\n';
            return output;
        }

        /**
         * @param {object} kw
         * @return {Keyframe}
         */

    }, {
        key: 'optimize',
        value: function optimize(kw) {
            this.stop = optimization.optimizeList(this.stop, kw);
            this.content = optimization.optimizeDeclarations(this.content, kw);
            return this;
        }
    }]);

    return Keyframe;
}();
