// crass/dist/nodes/Namespace.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

module.exports = function () {
    /**
     * @constructor
     * @param {URL} namespaceURI
     * @param {string} prefix
     */
    function Namespace(namespaceURI, prefix) {
        _classCallCheck(this, Namespace);

        this.namespaceURI = namespaceURI;
        this.prefix = prefix;
    }

    /**
     * @return {string}
     */


    _createClass(Namespace, [{
        key: 'toString',
        value: function toString() {
            if (this.prefix) {
                return '@namespace ' + this.prefix + ' ' + this.namespaceURI.toString() + ';';
            } else {
                return '@namespace ' + this.namespaceURI.toString() + ';';
            }
        }

        /**
         * @return {string}
         */

    }, {
        key: 'pretty',
        value: function pretty() {
            return this.toString() + '\n';
        }

        /**
         * @return {Namespace}
         */

    }, {
        key: 'optimize',
        value: function optimize() {
            return this;
        }
    }]);

    return Namespace;
}();
