// crass/dist/nodes/NthSelector.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var objects = require('../objects');
var optimization = require('../optimization');

module.exports = function () {
    /**
     * @constructor
     * @param {string} funcName
     * @param {LinearFunction} linearFunc
     */
    function NthSelector(funcName, linearFunc) {
        _classCallCheck(this, NthSelector);

        this.funcName = funcName;
        this.linearFunc = linearFunc;
    }

    /**
     * @return {string}
     */


    _createClass(NthSelector, [{
        key: 'toString',
        value: function toString() {
            return ':' + this.funcName + '(' + this.linearFunc.toString() + ')';
        }

        /**
         * @param {int} indent
         * @return {string}
         */

    }, {
        key: 'pretty',
        value: function pretty(indent) {
            var lfPretty = this.linearFunc.pretty ? this.linearFunc.pretty(indent) : this.linearFunc.toString();
            return ':' + this.funcName + '(' + lfPretty + ')';
        }

        /**
         * @param {object} kw
         * @return {NthSelector}
         */

    }, {
        key: 'optimize',
        value: function optimize(kw) {
            this.linearFunc = optimization.try_(this.linearFunc, kw);

            // OPT: nth-selectors (2n+1) to (odd)
            if (this.linearFunc.toString() === '2n+1') {
                return new objects.NthSelector(this.funcName, 'odd');
            }

            return this;
        }
    }]);

    return NthSelector;
}();
