// crass/dist/nodes/PseudoSelectorFunction.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var optimization = require('../optimization');

module.exports = function () {
    /**
     * @constructor
     * @param {string} funcName
     * @param {Expression} expr
     */
    function PseudoSelectorFunction(funcName, expr) {
        _classCallCheck(this, PseudoSelectorFunction);

        this.funcName = funcName;
        this.expr = expr;
    }

    /**
     * @return {string}
     */


    _createClass(PseudoSelectorFunction, [{
        key: 'toString',
        value: function toString() {
            return ':' + this.funcName + '(' + this.expr.toString() + ')';
        }

        /**
         * @param {int} ident
         * @return {string}
         */

    }, {
        key: 'pretty',
        value: function pretty(indent) {
            return ':' + this.funcName + '(' + this.expr.pretty(indent) + ')';
        }

        /**
         * @param {object} kw
         * @return {PseudoSelectorFunction}
         */

    }, {
        key: 'optimize',
        value: function optimize(kw) {
            // OPT: Lowercase pseudo function names.
            this.funcName = this.funcName.toLowerCase();
            this.expr = optimization.try_(this.expr, kw);
            return this;
        }
    }]);

    return PseudoSelectorFunction;
}();
